tock/hours/models.py

```diff
@@ -236,8 +258,47 @@ def save(self, *args, **kwargs):
         """
         if not self.id and self.user:
             self.billable_expectation = self.user.user_data.billable_expectation
+        self.calculate_hours()
         super().save(*args, **kwargs)
 
+    def calculate_utilization_denominator(self):
+        return self.billable_hours + self.non_billable_hours
+
+    def calculate_target_hours(self):
+        return round(self.billable_expectation * self.calculate_utilization_denominator(), 0)
+
+    def calculate_utilization(self):
+        if self.target_hours:
+            return self.billable_hours / self.target_hours
+        return 0
+
+    def calculate_hours(self):
+        billable_filter = Q(timecardobjects__project__accounting_code__billable=True)
+        non_billable_filter = Q(timecardobjects__project__accounting_code__billable=False,
+                          timecardobjects__project__exclude_from_billability=False)
+        excluded_filter = Q(timecardobjects__project__exclude_from_billability=True)
+
+        # Using Coalesce to set a default value of 0 if no data is available
+        billable = Coalesce(Sum('timecardobjects__hours_spent', filter=billable_filter), 0)
+        non_billable = Coalesce(Sum('timecardobjects__hours_spent', filter=non_billable_filter), 0)
+        excluded = Coalesce(Sum('timecardobjects__hours_spent', filter=excluded_filter), 0)
+
+        timecard = Timecard.objects.filter(id=self.id).annotate(billable=billable).annotate(non_billable=non_billable).annotate(excluded=excluded)[0]
+
+        self.billable_hours = round(timecard.billable, 2)
+        self.non_billable_hours = round(timecard.non_billable, 2)
+        self.excluded_hours = round(timecard.excluded, 2)
+
+        self.target_hours = self.calculate_target_hours()
+        self.utilization = self.calculate_utilization()
+
+    def _update_utilization_fields(self):
```

Comment:
This one could maybe use a comment, and I think `calc_target_hours` above could just call this for less repetition

Reply:
Whoops, good catch. Vestigial and not being called anywhere -- removing altogether.